### PR TITLE
[libmpeg2] Add new port

### PR DESCRIPTION
**Pull request by janisozaur**
Libmpeg2 seems to be a somewhat old project, last release was done in
2008, but it is still in use by some projects e.g. ResidualVM

I have verified the mpeg2dec built for x86-64 with MSVC produces similar
results as the one pre-packaged on my Arch.

### CODE REVIEW

File: ports/libmpeg2/CMakeLists.txt [hunk 1]
```diff
@@ -108,4 +108,4 @@ install(TARGETS mpeg2
     LIBRARY DESTINATION lib
 )
 
-install(FILES ${HEADERS} DESTINATION "include/libmpeg2")
+install(FILES ${HEADERS} DESTINATION "include/mpeg2dec")
```
Comment (janisozaur) [inline ports/libmpeg2/CMakeLists.txt]: Evidence:
* https://packages.ubuntu.com/disco/amd64/libmpeg2-4-dev/filelist
* https://www.archlinux.org/packages/extra/x86_64/libmpeg2/files/

File: ports/libmpeg2/CONTROL [hunk 2]
```diff
@@ -0,0 +1,8 @@
+Source: libmpeg2
+Version: 0.5.1
+Homepage: http://libmpeg2.sourceforge.net/
+Description: a free MPEG-2 video stream decoder
+
+Feature: build_binaries
+Description: Build binaries provided with libmpeg2
+Build-Depends:
```
Comment (NancyLi1013) [inline ports/libmpeg2/CONTROL]: If there is no dependency, please remove it.

File: ports/libmpeg2/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,48 @@
+vcpkg_fail_port_install(MESSAGE "${PORT} currently only supports Windows platform." ON_TARGET "Linux" "OSX")
+if(VCPKG_CMAKE_SYSTEM_NAME STREQUAL WindowsStore)
+    message(FATAL_ERROR "Error: UWP build is not supported.")
+endif()
```
Comment (NancyLi1013) [inline ports/libmpeg2/portfile.cmake]: These lines can be merged with above code
```
if(VCPKG_CMAKE_SYSTEM_NAME STREQUAL WindowsStore)
    message(FATAL_ERROR "Error: UWP build is not supported.")
endif()
```
You can add `uwp` by `ON_TARGET` to `vcpkg_fail_port_install()`.

File: ports/libmpeg2/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS ${FEATURE_OPTIONS}
+)
+
+vcpkg_install_cmake()
+
+# # Moves all .cmake files from /debug/share/libmpeg2/ to /share/libmpeg2/
+# # See /docs/maintainers/vcpkg_fixup_cmake_targets.md for more details
+# vcpkg_fixup_cmake_targets(CONFIG_PATH cmake TARGET_PATH share/libmpeg2)
```
Comment (NancyLi1013) [inline ports/libmpeg2/portfile.cmake]: If these lines are no longer needed, please remove them.

File: ports/libmpeg2/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+# # Handle copyright
+file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/libmpeg2 RENAME copyright)
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+
+vcpkg_copy_pdbs()
+
+# # Post-build test for cmake libraries
+vcpkg_test_cmake(PACKAGE_NAME libmpeg2)
```
Comment (NancyLi1013) [inline ports/libmpeg2/portfile.cmake]: `vcpkg_test_cmake()` is already outdated. Please remove it.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
Feature tested pass with the following triplets:

- x86-windows   (pass)

- x64-windows-static (pass)

- arm64-windows (pass)

- arm-uwp (not supported)

- x64-uwp (not supported)

This port currently only supports windows platform except for `arm64-uwp, x86-uwp, x64-uwp` triplets.

**Comment by janisozaur** [comment]
Updated with all the requests

File: ports/libmpeg2/portfile.cmake (hunk 5 above)
Comment (janisozaur) [reply ports/libmpeg2/portfile.cmake]: done

File: ports/libmpeg2/portfile.cmake (hunk 4 above)
Comment (janisozaur) [reply ports/libmpeg2/portfile.cmake]: done

File: ports/libmpeg2/portfile.cmake (hunk 3 above)
Comment (janisozaur) [reply ports/libmpeg2/portfile.cmake]: done

File: ports/libmpeg2/CONTROL (hunk 2 above)
Comment (janisozaur) [reply ports/libmpeg2/CONTROL]: done

**Review by NancyLi1013 — APPROVED** [review]

**Comment by janisozaur** [comment]
Ping

**Comment by janisozaur** [comment]
Ping

**Comment by dan-shaw** [comment]
@janisozaur Currently the MacOS pipeline is down, so we will have to wait on this.

**Comment by janisozaur** [comment]
Is it required even if the package explicitly doesn't support that os? https://github.com/microsoft/vcpkg/blob/8e0d3e78284c6c8847ceea3dd1ad6f23c44997e9/ports/libmpeg2/portfile.cmake#L1

**Comment by NancyLi1013** [comment]
Hi @janisozaur thanks for this PR.
Since this port doesn't support osx, linux, uwp.
Could you please update [ci.baseline.txt](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) with the followings?

```
libmpeg2:x64-uwp=fail
libmpeg2:arm-uwp=fail
libmpeg2:x64-linux=fail
libmpeg2:x64-osx=fail
```

**Comment by janisozaur** [comment]
Done

File: ports/libmpeg2/0001-Add-naive-MSVC-support-to-sources.patch [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+     else if (accel & MPEG2_ACCEL_X86_3DNOW)
+diff --git a/libvo/video_out_dx.c b/libvo/video_out_dx.c
+index 36de68a..d41a5a6 100644
+--- a/libvo/video_out_dx.c
++++ b/libvo/video_out_dx.c
+@@ -37,6 +37,7 @@
+ 
+ #include <ddraw.h>
+ #include <initguid.h>
++//#include <windows.h>
```
Comment (ras0219-msft) [inline ports/libmpeg2/0001-Add-naive-MSVC-support-to-sources.patch]: Stray comment

File: ports/libmpeg2/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,111 @@
+cmake_minimum_required(VERSION 3.2)
+project(libmpeg2)
+
+option(BUILD_BINARIES "Build libmpeg2 binaries" OFF)
+if (BUILD_SHARED_LIBS)
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
Comment (ras0219-msft) [inline ports/libmpeg2/CMakeLists.txt]: Per policy, we do not apply `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`:
https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#do-not-add-cmake_windows_export_all_symbols

If the library doesn't have appropriate `__declspec(dllexport)` annotations (or macros), it should be marked static only

File: ports/libmpeg2/CONTROL [hunk 8]
```diff
@@ -0,0 +1,7 @@
+Source: libmpeg2
+Version: 0.5.1
+Homepage: http://libmpeg2.sourceforge.net/
+Description: a free MPEG-2 video stream decoder
+
+Feature: build_binaries
```
Comment (ras0219-msft) [inline ports/libmpeg2/CONTROL]: Conventionally, we have called features like this "tool" or "tools"

**Review by ras0219-msft — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
Hi @janisozaur 

Could you please address the review suggestions?

**Comment by janisozaur** [comment]
I will try spending some time on it this week.

**Comment by janisozaur** [comment]
Updated. I also created an archive repository of imported `libmpeg2` releases: https://github.com/janisozaur/libmpeg2, as I don't really trust sourceforge that much and their SVN/CVS is broken.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/libmpeg2/CONTROL [hunk 9]
```diff
@@ -0,0 +1,7 @@
+Source: libmpeg2
+Version: 0.5.1
+Homepage: http://libmpeg2.sourceforge.net/
+Description: a free MPEG-2 video stream decoder
+
```
[the remark this answers is not in the document]
Comment (janisozaur) [reply ports/libmpeg2/CONTROL]:
```suggestion
Supports: !(linux|osx|uwp)

```

[the remark this answers is not in the document]
Comment (janisozaur) [reply ports/libmpeg2/CONTROL]: Updated

**Comment by janisozaur** [comment]
The CI runs failed not because of this PR. I'm assuming the vcpkg team will handle that.

**Comment by NancyLi1013** [comment]
@janisozaur 
Thanks for your notice.
We will try to fix the regressions that are not related with this PR.
Once fixed, I will re run this PR.

**Comment by janisozaur** [comment]
Looks good to go now, or is there anything else?

**Comment by NancyLi1013** [comment]
LGTM. This needs to be further reviewed by our other colleagues.
Before this, there is nothing else to do now.
Thanks for your PR.

**Review by NancyLi1013 — APPROVED** [review]

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
LGTM, thanks for this PR!

**Comment by janisozaur** [comment]
[`##[error] REGRESSION: qt5-tools:x64-linux`](https://dev.azure.com/vcpkg/public/_build/results?buildId=32984&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=a4822bc1-f3a9-5984-c941-6c1b7aa2ab81&l=25)

Unrelated regression again.

**Comment by NancyLi1013** [comment]
Thanks for your attention.
We will fix the regression as soon as possible.

**Comment by janisozaur** [comment]
All checks passing now

**Comment by NancyLi1013** [comment]
Thanks for your reminder.
LGTM now.

**Review by ras0219-msft — APPROVED** [review]

**Comment by janisozaur** [comment]
Finally all green.

File: ports/libmpeg2/0002-Mark-functions-as-exported.patch [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+index aff0c33..3520cbe 100644
+--- a/include/mpeg2.h
++++ b/include/mpeg2.h
+@@ -41,6 +41,11 @@
+ #define SEQ_VIDEO_FORMAT_MAC 0x80
+ #define SEQ_VIDEO_FORMAT_UNSPECIFIED 0xa0
+ 
++// Set to desired linkage qualifier, e.g. __declspec(dllexport)
++#ifndef LIBMPEG2_API
++#define LIBMPEG2_API
```
Comment (strega-nil) [inline ports/libmpeg2/0002-Mark-functions-as-exported.patch:21]: Hey, this library seems to only support building statically; can you remove this patch?

**Review by strega-nil — CHANGES_REQUESTED** [review]
That's the only thing I see that needs changing; this is really cool, thanks!

File: ports/libmpeg2/0002-Mark-functions-as-exported.patch (hunk 10 above)
Comment (janisozaur) [reply ports/libmpeg2/0002-Mark-functions-as-exported.patch]: Done

**Comment by janisozaur** [comment]
Ping. It seems there is misconfiguration in the Azure Pipelines, this job has been running for close to half an hour now, in particular the `x64_linux` one, while this port only supports Windows and I'm sure it would take much less than this to compile vcpkg+libmpeg2 if it were supported. Previous force push also stalled for about two days.

**Comment by strega-nil** [comment]
@janisozaur this looks good; I'll merge as soon as we get LLVM working again.

**Comment by strega-nil** [comment]
Alright, awesome! Thanks @janisozaur, this is great 😄

**Comment by janisozaur** [comment]
Finally! Thanks, now on to updating that scummvm build…

**Pull request merged**